source/common/grpc/json_transcoder_filter.cc

```diff
@@ -0,0 +1,398 @@
+
+  const auto& request_status = transcoder_->RequestStatus();
+
+  if (!request_status.ok()) {
+    ENVOY_LOG(debug, "Transcoding request error " + request_status.ToString());
+    error_ = true;
+    Http::Utility::sendLocalReply(*decoder_callbacks_, Http::Code::BadRequest,
+                                  request_status.error_message().ToString());
+
+    return Http::FilterDataStatus::StopIterationNoBuffer;
+  }
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus JsonTranscoderFilter::decodeTrailers(Http::HeaderMap&) {
+  if (!transcoder_) {
+    return Http::FilterTrailersStatus::Continue;
+  }
+
+  request_in_.finish();
+
+  Buffer::OwnedImpl data;
+  readToBuffer(*transcoder_->RequestOutput(), data);
+
+  if (data.length()) {
+    decoder_callbacks_->addDecodedData(data);
+  }
+  return Http::FilterTrailersStatus::Continue;
+}
+
+void JsonTranscoderFilter::setDecoderFilterCallbacks(
+    Http::StreamDecoderFilterCallbacks& callbacks) {
+  decoder_callbacks_ = &callbacks;
+}
+
+Http::FilterHeadersStatus JsonTranscoderFilter::encodeHeaders(Http::HeaderMap& headers,
+                                                              bool end_stream) {
+  if (error_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (!transcoder_) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  response_headers_ = &headers;
+  headers.insertContentType().value(Http::Headers::get().ContentTypeValues.Json);
+  if (!method_->server_streaming() && !end_stream) {
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus JsonTranscoderFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (error_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  if (!transcoder_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  response_in_.move(data);
+
+  if (end_stream) {
+    response_in_.finish();
+  }
+
+  readToBuffer(*transcoder_->ResponseOutput(), data);
+
+  if (!method_->server_streaming()) {
+    return Http::FilterDataStatus::StopIterationAndBuffer;
+  }
+  // TODO(lizan): Check ResponseStatus
+
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus JsonTranscoderFilter::encodeTrailers(Http::HeaderMap& trailers) {
+  if (!transcoder_) {
+    return Http::FilterTrailersStatus::Continue;
+  }
+
+  response_in_.finish();
+
+  Buffer::OwnedImpl data;
+  readToBuffer(*transcoder_->ResponseOutput(), data);
+
+  if (data.length()) {
+    encoder_callbacks_->addEncodedData(data);
+  }
+
+  if (method_->server_streaming()) {
+    // For streaming case, the headers are already sent, so just continue here.
+    return Http::FilterTrailersStatus::Continue;
+  }
+
+  const Http::HeaderEntry* grpc_status_header = trailers.GrpcStatus();
+  if (grpc_status_header) {
+    uint64_t grpc_status_code;
+    if (!StringUtil::atoul(grpc_status_header->value().c_str(), grpc_status_code)) {
```

Comment:
Hmm, since we're just forward grpc-status to client, so I don't need the additional error checking here. There is a plan to have HTTP code mapping, will do that in follow up.